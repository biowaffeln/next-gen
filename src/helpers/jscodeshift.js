// @ts-check
import addImports from "jscodeshift-add-imports";

/**
 * @typedef {import("jscodeshift").JSCodeshift} JSCodeshift
 * @typedef {import("jscodeshift").Collection} Collection
 * @typedef {import("jscodeshift").ObjectProperty} ObjectProperty
 */

/**
 * Adds an import statement to a .js or .ts file.
 * Merges imports if they exist, otherwise it appends
 * the import after the existing ones.
 *
 * Similar to `jscodeshift-add-imports`, but fixes some
 * formatting issues regarding blank lines.
 *
 * @param {JSCodeshift} j
 * @param {Collection} root
 * @param {string} statement
 */
export function addImport(j, root, statement) {
	const s = j.template.statement([statement]);

	const imports = root.find(j.ImportDeclaration);
	const newImport = j(s);

	// only merge when the sources and importKinds are the same
	const shouldMerge =
		imports.filter(
			(p) =>
				p.get("source").node.value === newImport.get("source").node.value &&
				p.get().node.importKind === newImport.get().node.importKind
		).length !== 0;

	// if there are no imports, add statement with two line breaks
	if (imports.length === 0) {
		const s = j.template.statement([statement + "\n\n"]);
		root.get().node.program.body.unshift(s);
	}
	// if there are imports but no merging is required,
	// add the statement at the end of the imports (without blank lines inbetween)
	else if (!shouldMerge) {
		imports.at(-1).replaceWith((p) =>
			j.template.statements([
				`${j(p).toSource()}
${statement}`,
			])
		);
	}
	// otherwise, merge imports
	else {
		addImports(root, [s]);
	}
}

/**
 * Adds a require statement to a .js file.
 * Does not merge existing require calls.
 *
 * @param {JSCodeshift} j
 * @param {Collection} root
 * @param {string} statement
 */
export function addRequire(j, root, statement) {
	// find variable declarations that call require
	const declarations = root.find(j.VariableDeclaration, {
		declarations: {
			0: { init: { callee: { name: "require" } } },
		},
	});

	// if there aren't any, add require to top of the file
	if (declarations.length === 0) {
		const s = j.template.statement([statement + "\n\n"]);
		root.get().node.program.body.unshift(s);
	}
	// otherwise add it after last require
	else {
		declarations.at(-1).replaceWith((p) =>
			j.template.statements([
				`${j(p).toSource()}
	${statement};`,
			])
		);
	}
}

/**
 *
 * @param {JSCodeshift} j
 * @param {Collection} root
 * @param  {ObjectProperty[]} objectProperties
 */
export function addWebpackOptions(j, root, ...objectProperties) {
	root
		.find(j.ObjectExpression)
		.at(-1)
		.forEach((p) => p.node.properties.push(...objectProperties));
}
