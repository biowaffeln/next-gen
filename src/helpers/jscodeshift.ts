import type { Collection, JSCodeshift, ObjectProperty } from "jscodeshift";
import type { ExpressionKind } from "ast-types/gen/kinds";
import addImports from "jscodeshift-add-imports";

/**
 * Adds an import statement to a .js or .ts file.
 * Merges imports if they exist, otherwise it appends
 * the import after the existing ones.
 *
 * Similar to `jscodeshift-add-imports`, but fixes some
 * formatting issues regarding blank lines.
 */
export function addImport(j: JSCodeshift, root: Collection, statement: string) {
	const s = j.template.statement([statement]);

	const imports = root.find(j.ImportDeclaration);
	const newImport = j(s);

	// only merge when the sources and importKinds are the same
	const shouldMerge =
		imports.filter(
			(p) =>
				p.get("source").node.value === newImport.get("source").node.value &&
				p.get().node.importKind === newImport.get().node.importKind
		).length !== 0;

	// if there are no imports, add statement with two line breaks
	if (imports.length === 0) {
		const s = j.template.statement([statement + "\n\n"]);
		root.get().node.program.body.unshift(s);
	}
	// if there are imports but no merging is required,
	// add the statement at the end of the imports (without blank lines inbetween)
	else if (!shouldMerge) {
		imports.at(-1).replaceWith((p) =>
			j.template.statements([
				`${j(p).toSource()}
${statement}`,
			])
		);
	}
	// otherwise, merge imports
	else {
		addImports(root, [s]);
	}
}

/**
 * Adds a require statement to a .js file.
 * Does not merge existing require calls.
 */
export function addRequire(
	j: JSCodeshift,
	root: Collection,
	statement: string
) {
	// find variable declarations that call require
	const declarations = root.find(j.VariableDeclaration, {
		declarations: {
			0: { init: { callee: { name: "require" } } },
		},
	});

	// if there aren't any, add require to top of the file
	if (declarations.length === 0) {
		const s = j.template.statement([statement + "\n\n"]);
		root.get().node.program.body.unshift(s);
	}
	// otherwise add it after last require
	else {
		declarations.at(-1).replaceWith((p) =>
			j.template.statements([
				`${j(p).toSource()}
	${statement};`,
			])
		);
	}
}

export function wrapNextConfig(
	j: JSCodeshift,
	root: Collection,
	wrapper: ExpressionKind
) {
	const moduleExport = root.find(j.AssignmentExpression, {
		left: {
			object: { name: "module" },
			property: { name: "exports" },
		},
	});

	if (moduleExport.length === 0) {
		throw new Error("No module export was found.");
	}

	moduleExport.forEach((p) => {
		p.node.right = j.callExpression(wrapper, [p.node.right]);
	});
}

export function addConfigOptions(
	j: JSCodeshift,
	root: Collection,
	...objectProperties: ObjectProperty[]
) {
	const moduleExport = root.find(j.AssignmentExpression, {
		left: {
			object: { name: "module" },
			property: { name: "exports" },
		},
	});

	if (moduleExport.length === 0) {
		throw new Error("Failed to add nextConfig properties.");
	}

	const right = moduleExport.get("right");

	// simple case
	if (right.node.type === "ObjectExpression") {
		right.node.properties.push(...objectProperties);
	}
	// complex case
	else if (right.node.type === "CallExpression") {
		const composePlugin = root.find(j.VariableDeclarator, {
			init: {
				callee: { name: "require" },
				arguments: { 0: { value: "next-compose-plugins" } },
			},
		});

		// error if next config requires `next-compose-plugins`
		// TODO: fix
		if (composePlugin.length > 0) {
			throw new Error("Faild to add nextConfig properties.");
		}

		// find the innermost call expression
		moduleExport
			.find(j.CallExpression, (value) => {
				return j(value).find(j.CallExpression).length === 0;
			})
			// add objectExpression or insert into existing one
			.forEach((p) => {
				if (
					p.node.arguments[0] &&
					p.node.arguments[0].type === "ObjectExpression"
				) {
					p.node.arguments[0].properties.push(...objectProperties);
				} else {
					p.node.arguments.push(j.objectExpression(objectProperties));
				}
			});
	} else {
		throw new Error("Failed to add nextConfig properties.");
	}
}
