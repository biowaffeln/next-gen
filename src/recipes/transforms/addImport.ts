import * as j from 'jscodeshift';

export function addImport(
	program: j.Collection<j.Program>,
	importToAdd: j.ImportDeclaration
): j.Collection<j.Program> {
	const imports = program.find(j.ImportDeclaration);
	if (imports.length === 0) {
		program.find(j.Statement).at(0).insertBefore(importToAdd);
	} else {
		imports.at(-1).insertAfter(importToAdd);
	}
	return program;
}
