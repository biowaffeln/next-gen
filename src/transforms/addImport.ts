import * as j from 'jscodeshift';

export const addImport = (
	program: j.Collection<j.Program>,
	importToAdd: j.ImportDeclaration
): j.Collection<j.Program> => {
	const imports = program.find(j.ImportDeclaration);
	if (imports.length === 0) {
		program.get().node.program.body.unshift(importToAdd, ' ');
	} else {
		imports.at(-1).insertAfter(importToAdd);
	}
	return program;
};
