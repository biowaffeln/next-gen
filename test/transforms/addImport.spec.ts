import * as j from 'jscodeshift';
import { addImport } from '../../src/transforms/addImport';

const p = j.withParser('tsx');

const emptyProgram = p('');
const programWithStatement = p('export const test = 123;');
const programWithImports = p(`import { render } from 'react-dom';

render(<div />, document.getElementById('root'));`);

const importToAdd = j.template.statement`import { foo } from "bar";`;

test('should add imports to empty file', () => {
	expect(addImport(emptyProgram, importToAdd).toSource()).toMatchSnapshot();
});

test('should add imports to file with content', () => {
	expect(
		addImport(programWithStatement, importToAdd).toSource()
	).toMatchSnapshot();
});

test('should add imports to file with existing imports', () => {
	expect(
		addImport(programWithImports, importToAdd).toSource()
	).toMatchSnapshot();
});
