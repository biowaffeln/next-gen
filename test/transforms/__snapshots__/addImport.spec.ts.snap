// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add imports to empty file 1`] = `
"import { foo } from \\"bar\\";
 "
`;

exports[`should add imports to file with content 1`] = `
"import { foo } from \\"bar\\";
 
export const test = 123;"
`;

exports[`should add imports to file with existing imports 1`] = `
"import { render } from 'react-dom';

import { foo } from \\"bar\\";

render(<div />, document.getElementById('root'));"
`;
